File: training-data/prolog/longest_increasing_subsequence.pl
subsequence([], []).
subsequence([H|T], [H|Rest]) :- subsequence(T, Rest).
subsequence([_|T], Rest) :- subsequence(T, Rest).

is_increasing([_]).
is_increasing([X, Y|Rest]) :- X < Y, is_increasing([Y|Rest]).

all_subsequence_lengths(List, Lengths) :-
    findall(SubLen,
            ( subsequence(List, Sub), Sub \= [], is_increasing(Sub), length(Sub, SubLen) ),
            Lengths).

longest_increasing_subsequence(List, Best) :-
    all_subsequence_lengths(List, Lengths),
    max_list(Lengths, Best).

:- longest_increasing_subsequence([10, 9, 2, 5, 3, 7, 101, 18], Best),
   writeln(Best).

File: training-data/pascal/longest_increasing_subsequence.pas
program LongestIncreasingSubsequenceDemo;

const
  N = 8;

var
  data: array[0..N - 1] of Integer = (10, 9, 2, 5, 3, 7, 101, 18);
  lengths: array[0..N - 1] of Integer;
  i, j, best: Integer;

begin
  for i := 0 to N - 1 do
    lengths[i] := 1;

  for i := 1 to N - 1 do
    for j := 0 to i - 1 do
      if (data[j] < data[i]) and (lengths[j] + 1 > lengths[i]) then
        lengths[i] := lengths[j] + 1;

  best := 0;
  for i := 0 to N - 1 do
    if lengths[i] > best then best := lengths[i];

  WriteLn('longest increasing subsequence length: ', best);
end.
